### ISSUE TITLE
RFM 论文讨论

### ISSUE THREAD

**Issue by slsongge** [JiaxiangBU/tutoring#15]
![image](https://user-images.githubusercontent.com/36298417/56465044-5dff6980-6428-11e9-929f-9353d6df7c1b.png)
上图是其中一篇比较好的论文的思路，个人感觉挺完美的，不过我的论文肯定得有不同之处。
1、我在聚类的地方想增加点内容，比如说加一个类似的聚类方法（可以指定簇的个数&尽量不基于距离）。
2、顺便加一个可以进行对比两者聚类效果的方法（我试图使用clustree，疑问是：是不是只能额外的加入一个主成分去判断其效果，如果是的话，那又回归到那个这个主成分尽可能对两个聚类是无关的，这样保持相对公平）
3、论文中涉及到了利用AHP进行更加精准的营销的方法，我感觉我也用这个方法就有些不妥。其实我现在对AHP和结构方程几乎还是零理解，如果你感觉这个有必要（或者可以出现创新点），我就尝试一下
4、后面的分类问题其实就无所谓了，不过也有不理解之处：他自定义了回头率（客户第二年乘机次数与第一年乘机次数比值，0.5, 0.8作为阈值）。我是否可以更加精细化一下这个定义，比如说我添加一下客户“活动规律“（比如某些人比值小，但是出行很有规律，我就可以重新把这部分人更好的进行定义标签）、起始地-到达地（尝试挖掘些信息去更加精细的定义标签）等维度去更加精细的定义标签。
5、就是分类问题中，我始终不是很清楚随即森林能不能做到类似决策树那样规则可视化的效果，如果行的话我就使用随机森林。
6、辛苦家翔哥，无疑言谢！！！

**Comment by JiaxiangBU** [comment]
@slsongge 

这篇文献我看到是邮件附件中的。

![image](https://user-images.githubusercontent.com/15884785/56472772-5b872900-6495-11e9-90ec-6834fdd937b7.png)

这个地方我觉得有些问题，我觉得没有必要算完了RFM，聚类了32个组，又合并，这是**信息损失的操作**。不如，直接合并好小组。因此我去看了他的具体思路。

![image](https://user-images.githubusercontent.com/15884785/56472758-3c889700-6495-11e9-9677-b28d5a5497ec.png)

这个地方的思路我觉得没有问题，K-Means 针对这种维度比较少的情况效果是比较好的。

- [ ] 这里我可以找一下“K-Means 针对这种维度比较少的情况效果是比较好的”的笔记

因此我的感觉是这里你基于五个变量的特征变量去选择某种方法的提升应该不大。

> 1、我在聚类的地方想增加点内容，比如说加一个类似的聚类方法（可以指定簇的个数&尽量不基于距离）。

**这里你除非考虑把特征变量的原始字段拿出来聚类**。
但是这样你的思路就有点不是很**优美**，过于复杂了，你可以需要考虑下。

方法上，我建议就是**结构方程**完事，因为这是这种因子关系，比较统一的方法，参考 #13 

> 2、顺便加一个可以进行对比两者聚类效果的方法（我试图使用clustree，疑问是：是不是只能额外的加入一个主成分去判断其效果，如果是的话，那又回归到那个这个主成分尽可能对两个聚类是无关的，这样保持相对公平）

不需要加入 PCA，clustree 的效果是真实样本的分类，你可以注意下图的计算过程。

> 3、论文中涉及到了利用AHP进行更加精准的营销的方法，我感觉我也用这个方法就有些不妥。其实我现在对AHP和结构方程几乎还是零理解，如果你感觉这个有必要（或者可以出现创新点），我就尝试一下

是的，结构方程是一个目前来说做这种小数据研究比较流行方式。

> 4、后面的分类问题其实就无所谓了，不过也有不理解之处：他自定义了回头率（客户第二年乘机次数与第一年乘机次数比值，0.5, 0.8作为阈值）。我是否可以更加精细化一下这个定义，比如说我添加一下客户“活动规律“（比如某些人比值小，但是出行很有规律，我就可以重新把这部分人更好的进行定义标签）、起始地-到达地（尝试挖掘些信息去更加精细的定义标签）等维度去更加精细的定义标签。

可以，但是我觉得太偏商业分析，没有算法(是模型不是计算口径)的优势。

> 5、就是分类问题中，我始终不是很清楚随即森林能不能做到类似决策树那样规则可视化的效果，如果行的话我就使用随机森林。

可以。

- [ ] 我回头找一个例子给你，是 Xgboost 的

**Comment by slsongge** [comment]
@JiaxiangBU 
两个问题：clustree和分类
1、我感觉下图的clustree不能够表达出聚类的效果，因为图中（下图）没有变量的参与，比如那个PC1的图（存在一个变量去衡量后面的类是不是已经开始重合，不能清晰的聚类了），若理解不对，请指正。
![image](https://user-images.githubusercontent.com/36298417/56482011-6e7d1600-64f4-11e9-9d57-4e8620e9cee7.png)
2、分类哪里，我的想法是偏向商业，我也正是想做一个偏商业。算法我就打算一带而过，还有最好能可视化最后的规则策略，这样更形象化。至于xbg的话，我并不是很希望用，因为太高大上了，驾驭不了。

**Comment by JiaxiangBU** [comment]
> 我感觉下图的clustree不能够表达出聚类的效果，因为图中（下图）没有变量的参与

这个说法我觉得不对。如图，每个K的标签，如K=2，反馈的是一个唯一值为1或者是2的序列，对么？
这个序列是无监督学习产生的？无监督学习产生的逻辑依然是通过特征变量提取的，因此这个序列代表了特征变量。



> 存在一个变量去衡量后面的类是不是已经开始重合，不能清晰的聚类了

这个我没看懂。

> 至于xbg的话，我并不是很希望用，因为太高大上了，驾驭不了。

这个地方目的不是用 Xgboost，而是去学习它可视化的逻辑和代码，因此可以用 rf 和 Xgboost，构建都是靠 digrammR 包完成的。

**Comment by slsongge** [comment]
@JiaxiangBU 
今晚方便给你打一个电话吗？

> 这个说法我觉得不对。如图，每个K的标签，如K=2，反馈的是一个唯一值为1或者是2的序列，对么？
这个序列是无监督学习产生的？无监督学习产生的逻辑依然是通过特征变量提取的，因此这个序列代表了特征变量。

可能我对clustree理解不够深，或者你没有理解我的意思。举个例子哈，不一定合适，但是应该能让你更加理解我的意思。如下图，图中横轴是PC1（假设PC1没参与聚类），分别是k-mean和k-median聚类的结果。那么我就可以通过这两个clustree的节点交叉度得出k-median效果更好。不知道我这次解释的清楚吗？
![image](https://user-images.githubusercontent.com/36298417/56575257-e08a5380-65f7-11e9-896b-9a7addf3a241.png)

> 这个地方目的不是用 Xgboost，而是去学习它可视化的逻辑和代码，因此可以用 rf 和 Xgboost，构建都是靠 digrammR 包完成的。

rf的原理我比较清楚点，但是digrammR没有接触过，我今晚看看。

> 有关聚成32类的问题

我很好奇他怎么做到聚成32类的（并且每类的各个指标大小组合均不一致），如下图：
![image](https://user-images.githubusercontent.com/36298417/56577285-357c9880-65fd-11e9-9ef9-feb6d53daf5f.png)

> 有关结构方程的问题

我想问一下，结构方程是可以利用这5个指标计算出各自的权重吗？我怎么感觉不可思议啊。。。

**Comment by JiaxiangBU** [comment]
> 如下图，图中横轴是PC1（假设PC1没参与聚类），分别是k-mean和k-median聚类的结果。那么我就可以通过这两个clustree的节点交叉度得出k-median效果更好。不知道我这次解释的清楚吗？

去掉 PC1 呢？

**Comment by JiaxiangBU** [comment]
> 我很好奇他怎么做到聚成32类的（并且每类的各个指标大小组合均不一致），如下图：

这部分我觉得不用参考，他的报告我看过，因为是不能重复试验出来的。

**Comment by JiaxiangBU** [comment]
> 我想问一下，结构方程是可以利用这5个指标计算出各自的权重吗？我怎么感觉不可思议啊。。。

你看我之前给的资料。

**Comment by slsongge** [comment]
@JiaxiangBU 
> 去掉 PC1 呢？

去掉PC1就会导致树的节点不会交叉重叠吧？如果不交叉重叠，那怎么能看出两个聚类方法的好坏了呢？

**Comment by JiaxiangBU** [comment]
> 去掉PC1就会导致树的节点不会交叉重叠吧？如果不交叉重叠，那怎么能看出两个聚类方法的好坏了呢？

你可以整体看下基本上每个 K 均匀分配的，效果是比较好的。

统计方面，比较常用的方式你可以参考

* [clustering - Performance metrics to evaluate unsupervised learning - Cross Validated](https://stats.stackexchange.com/questions/79028/performance-metrics-to-evaluate-unsupervised-learning)
* [unsupervised learning - Evaluation measures of goodness or validity of clustering (without having truth labels) - Cross Validated](https://stats.stackexchange.com/questions/21807/evaluation-measures-of-goodness-or-validity-of-clustering-without-having-truth)

另外，如果你的算法不仅可以 fit 还可以 predict，那么就可以用交叉验证 cross validation 进行验证。

**Comment by slsongge** [comment]
> > 我想问一下，结构方程是可以利用这5个指标计算出各自的权重吗？我怎么感觉不可思议啊。。。
> 
> 你看我之前给的资料 #13

原理我是没有看懂呢，不过我有一个疑问：下图中红框里面的就是每个自变量对应的权重吗？相当于AHP似的。
![image](https://user-images.githubusercontent.com/36298417/56636363-13d1ee80-669b-11e9-9c45-34d3967bf632.png)

**Comment by JiaxiangBU** [comment]
> 下图中红框里面的就是每个自变量对应的权重吗

类似于逻辑回归的 beta。

**Comment by slsongge** [comment]
> 类似于逻辑回归的 beta。

结构方程就是因子分析的一类吧，或者说很类似吧，我去学习一下因子分析。

**Comment by JiaxiangBU** [comment]
> 结构方程就是因子分析的一类吧

对的，嗯，我之前发过一些材料给你，R这方面包 psych 和 sem 都很不错。

**Comment by slsongge** [comment]
> 对的，嗯，我之前发过一些材料给你，R这方面包 psych 和 sem 都很不错。

那么，如果保留一个公因子，那么是不是就相当于获得各个指标的权重了。我看你用了cfa函数：
wisc4.fit <- cfa(model=wisc4.model, sample.cov=wisc4.cov, sample.nobs=550,std.lv=FALSE)
wisc4.fit %>% tidy %>% select(term,std.all)
![image](https://user-images.githubusercontent.com/36298417/56711640-31b45780-675e-11e9-9a7f-7bde89738244.png)

这返回的权重直接对各个指标加权求和那么就是综合得分了是吗？

**Comment by JiaxiangBU** [comment]
> 那么，如果保留一个公因子，那么是不是就相当于获得各个指标的权重了。我看你用了cfa函数：

从你给的图片里面的结果上看，是的。

你可以看一下这部分 [*2.1.2 Multi-factor Model*](https://jiaxiangli.netlify.com/2018/10/21/factor-analysis/#multi-factor-model)，是里面 loading 的解释，非常类似于线性回归。

> 这返回的权重直接对各个指标加权求和那么就是综合得分了是吗？


是的。

**Comment by slsongge** [comment]
> 是的。

那我唯一不懂的就是不知道因子载荷矩阵怎么计算出来的了，其实也就是计算过程，没办法，太笨了。你也不用给给我写计算过程哈，太费时间了。

**Comment by slsongge** [comment]
有关因子分析的小疑问：我看因子分析中构建因子变量的方法有一个是主成分分析法，那么这个不就完全是主成分分析了吗？如下图：
![image](https://user-images.githubusercontent.com/36298417/56726049-9d0f2100-6780-11e9-85d5-2dfa7d7ff068.png)
该图的链接：https://blog.csdn.net/ITzym/article/details/70140429

**Comment by JiaxiangBU** [comment]
> 有关因子分析的小疑问：我看因子分析中构建因子变量的方法有一个是主成分分析法，那么这个不就完全是主成分分析了吗？如下图：

在单层的，也就是最简单的因子分析情况下，可以用PCA，但是不代表这就是PCA。

**Comment by JiaxiangBU** [comment]
> 那我唯一不懂的就是不知道因子载荷矩阵怎么计算出来的了，其实也就是计算过程，没办法，太笨了。你也不用给给我写计算过程哈，太费时间了。

嗯，这部分你可以理解下，我觉得不是很复杂，基本上就是线性代数的矩阵运算。

**Comment by slsongge** [comment]
> 统计方面，比较常用的方式你可以参考
> * [clustering - Performance metrics to evaluate unsupervised learning - Cross Validated](https://stats.stackexchange.com/questions/79028/performance-metrics-to-evaluate-unsupervised-learning)

![image](https://user-images.githubusercontent.com/36298417/56972000-d12d7c00-6b9c-11e9-8b87-e56e73360ec5.png)

这个方法（上面的group truth）是不是得有因变量啊，我看(http://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation)
里面的好像大都是有因变量才能使用的方法啊？

**Comment by JiaxiangBU** [comment]
@slsongge 
所以你引用的这个方法 *ARI*，这个联机也提到了缺点

> Contrary to inertia, **ARI requires knowledge of the ground truth classes** while is almost never available in practice or requires manual assignment by human annotators (as in the supervised learning setting).
https://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation

**Comment by JiaxiangBU** [comment]
@slsongge 

你可以参考这个回复的下半部分

> If your unsupervised learning method is probabilistic, another option is to evaluate some probability measure (log-likelihood, perplexity, etc) on held out data.
https://stats.stackexchange.com/questions/79028/performance-metrics-to-evaluate-unsupervised-learning

这个我之前了解过，你可以参考这个
https://jiaxiangbu.github.io/learn_nlp/datacamp/topic-modeling-in-r/ch3-log-likelihood.html

- [ ] 但是我觉得这个更好的方式是能在 Cross Validation 交叉验证层面，会更理想。我还没弄出来，你有兴趣也可以一起想想。

**Comment by JiaxiangBU** [comment]
另外我有一个比较简单的思路，但是不知道有没有文献支持。

比如你有1000个样本，做 cv = 2 的方式，随机排序一下。

- [ ] index = 1:500，作为训练集，预测 index = 1:1000 的样本，给分类值
- [ ] index = 501:1000，作为训练集，预测 index = 1:1000 的样本，给分类值

如果这两个系统给出的分类值差异越小，那么说明模型在这个数据集上越稳定。

你觉得呢?

- [ ] 但是实现上，我还没想好。

**Comment by slsongge** [comment]
@JiaxiangBU 
家翔哥，这样吧，之前都是在纯理论的讨论。我这一两天就开始着手去做有关聚类的这部分，主要尝试kmeans和DBSCAN，然后去尝试对比效果、定义聚类结果了之类的。我试试到底如何哈哈。还有就是这个项目可以设为不公开吗？如果可以且方便的话，就设为不公开吧，你看如何？

**Comment by JiaxiangBU** [comment]
> 还有就是这个项目可以设为不公开吗？

这个项目，本身还有其他人的内容在里面因此是公开的。如果后面你的数据、代码比较敏感的话，我重新开一个项目即可。
